src/tokens/TokenImport.tsx

```diff
@@ -27,57 +30,149 @@ import variables from 'src/styles/variables'
+      ValoraAnalytics.track(AssetsEvents.import_token_already_supported, {
+        networkId,
+        tokenId,
+        tokenAddress,
+      })
+      return false
     }
-  }
 
-  const handleAddressFocus = () => {
-    setIsCompleteAddress(false)
     setError(null)
+    return true
+  }
+
+  const fetchTokenDetails = async (
+    tokenAddress: Address,
+    walletAddress: Address
+  ): Promise<TokenDetails> => {
+    const contract = getContract({
+      abi: erc20.abi,
+      address: tokenAddress,
+      publicClient: publicClient[networkIdToNetwork[networkId]],
+    })
+
+    const [symbol, decimals, name, balance] = await Promise.all([
+      contract.read.symbol(),
+      contract.read.decimals(),
+      contract.read.name(),
+      contract.read.balanceOf([walletAddress]),
+    ])
+
+    return { address: tokenAddress, symbol, decimals, name, balance }
   }
 
+  const validateContract = useAsyncCallback(
+    async (address: Address): Promise<TokenDetails> => {
+      if (!walletAddress || !isAddress(walletAddress)) {
+        // should never happen
+        throw new Error('No wallet address found when fetching token details')
+      }
+
+      const details = fetchTokenDetails(address, walletAddress)
+      const timeout = new Promise<never>((_, reject) =>
+        setTimeout(reject, FETCH_TOKEN_DETAILS_TIMEOUT_MS, new Error('Timeout'))
+      )
+
+      return Promise.race([timeout, details])
+    },
+    {
+      onSuccess: (details) => {
+        Logger.info(
+          TAG,
+          `Wallet ${walletAddress} holds ${formatUnits(details.balance, details.decimals)} ${
+            details.symbol
+          } (${details.name} = ${details.address})})`
+        )
+        setTokenSymbol(details.symbol)
+      },
+      onError: (error) => {
+        Logger.error(TAG, 'Could not fetch token details', error)
+        const tokenId = getTokenId(networkId, tokenAddress.toLowerCase())
+        if (error.message === 'Timeout') {
+          setError(t('tokenImport.error.timeout'))
+          ValoraAnalytics.track(AssetsEvents.import_token_timeout, {
+            networkId,
+            tokenId,
+            tokenAddress,
+          })
+        } else {
+          setError(t('tokenImport.error.notErc20Token'))
+          ValoraAnalytics.track(AssetsEvents.import_token_not_erc20, {
+            networkId,
+            tokenId,
+            tokenAddress,
+          })
+        }
+      },
+    }
+  )
+
   const ensure0xPrefixOrEmpty = (address: string) =>
     !address || address.startsWith('0x') ? address : `0x${address}`
 
-  const handleAddressBlur = () => {
+  const handleAddressBlur = async () => {
+    if (validateContract.status !== 'not-requested') {
+      return
+    }
     const address = ensure0xPrefixOrEmpty(tokenAddress)
     setTokenAddress(address)
-    validateAddress(address)
+    if (validateAddress(address)) {
+      await validateContract.execute(address).catch(() => undefined)
```

Comment:
Hmm, I'm not sure but the `.catch` looks weird to me. I guess we are handling the error directly in the `onError`. But if we don't care about handling the error here, can we remove the `.catch` block?

Reply:
ah, i had the same thought. i came across something similar when i was working on revoke phone number https://github.com/valora-inc/wallet/blob/263144a9cf2f126bd448aa3867f3d5ec203461d1/src/RevokePhoneNumber.tsx#L61-L67 - i am guessing that this catch block is serving the same purpose. perhaps it is better to be explicit and add a comment for the catch block